### PR TITLE
[REVIEW] fix dataframe.query returning null rows erroneously.

### PR DESCRIPTION
**Pull request by cwharris**
Addresses #1994, but not entirely, hence discussion.

By applying a pessimistic nullmask (any input null, output is null) to the output of `df.query(...)`, we can get a _close_ match to Pandas `df.query(...)` . Unforunately, it doesn't work when `!=` or `or` is introduced in certain circumstances. For instance, `None != None`  should evaluate True, but it won't. `True or None` should evaluate True, but it won't. The approach taken in this PR cannot fix those cases.

Should we address those cases? In my opinion, yes. However, that requires a more comprehensive approach. Specifically, I think we would need use `query`'s AST to generate a complementary nullmask AST. For starters, that would refactor any value comparison in to `a & b` and run against the validity bitmaps along side the original value query.(`!=` comparisons would need to be transformed to `!(a & b)`.

Afaict, we are already parsing the query AST in `queryutils.py`.

What do we think of the proposed PR, and are there other/better ways to solve this?

### CODE REVIEW

**Comment by shwina** [comment]
Initial thoughts:

Since Pandas apparently treats `None` the same as `nan` when doing comparisons, I wonder if we couldn't match Pandas behaviour just by doing a `fillna(nan)` prior to the query.

For integers, perhaps we could cast to float and then cast back? Of course this is expensive: so maybe we should default to *not* matching Pandas None/nan behaviour unless the user explicitly asks for it?

**Comment by cwharris** [comment]
IMO it's important that we match Pandas behavior wherever we reasonably can, and we otherwise don't have a good reason to deviate. As for converting to `fp` and doing `fillna` before we execute the query, this doesn't hold up for discreet values such as booleans or categorical data (strings too, I guess, but that's a whole other story I think?).

@shwina how would you feel if I implemented `pandas.eval` with null / validity support and used that in the implementation `query`? According to the [documentation](https://pandas.pydata.org/pandas-docs/version/0.25/reference/api/pandas.DataFrame.query.html), this is what Pandas does.

**Comment by shwina** [comment]
Null semantics are an example of something we can expect to deviate from Pandas' behaviour, since we don't use `nan` to indicate missing values (while Pandas does). This `a == None` and `a == nan` mean two completely different things in `cudf`, while they are equivalent in Pandas to my knowledge.

I believe casting bools to floats will maintain the same behaviour for `query`? Categoricals and strings are definitely another story, and I don't think we support them right now anyway.

Can `cudf.eval` be built on top the existing `queryutils.py`?

**Comment by cwharris** [comment]
My impression is that `queryutils` is `eval` with a narrow focus on `df.query`. I imagine one would build on the other, or they would be consolidated.

From a computer science standpoint, I agree that `a == None` would be true only if `a` were `None`. But in data science, we're not dealing with pointers, but instead missing data. A person cannot have a null age - we simply do not know the age, but it does indeed exist. This has implications for comparison operations, where `age < 18` would evaluate `True` if age were `None` and we treated it as a zero (which is what we do now).

I'd also speculate that there's no hard guarantee that an invalid value (behind a nullmask) is actually initialized properly, so if we don't consider the validity bit explicitly, `age > 18` could evaluate `True` even though `age` is null (because the trash value is > 18, and we don't consider the validity bit).

IMO, we _need_ a way to do missing value comparisons properly (like with NaN), and it would be convenient to have alternative operators (such as `is`) available to do "pointer" comparison. I think this is a core concept in data science

**Comment by jrhemstad** [comment]
I agree with @shwina that Pandas' null behavior is not something we should strive to match, and should instead do something more sensible when feasible. Pandas' null behavior is oftentimes bananas. We've made this decision several times to deviate from Pandas null-behavior. Most recently I can think of is that a groupby sum aggregation on all null values is `null` as opposed to `0`. 

> I'd also speculate that there's no hard guarantee that an invalid value (behind a nullmask) is actually initialized properly

No need to speculate. The value behind a "null" value is completely undefined. 

> I think this is a core concept in data science

That's just the thing, there seems to be no universal consensus among data science folks about how null values are handled. Pandas vs SQL can differ wildly in this regard, which just illustrates that Pandas is certainly not the universal Truth.

For example, in SQL, I believe that in your `age < 18` example would return `NULL` if `age == NULL`.

**Comment by shwina** [comment]
If it's possible for us to match Pandas' `query` behaviour with just the `fillna()` approach above - and if we decide to match Pandas behaviour at all - then I'd suggest using that to address the issue in the short term. My suggestion would be to use a `nan_as_null=False` parameter to toggle Pandas compatibility

Generalization of `queryutils` to `eval` can be done later and we can plug that back in to `query` when it's ready.

Does that sound alright?

**Comment by cwharris** [comment]
While I think it's important to consider how we effect users who migrate from Pandas, I don't think Pandas should be held as the standard by which cuDF is judged, so in the future I'm going to make an effort to avoid verbiage that indicates otherwise. So far I have failed to do this.

So, being more specific and to the point, what would be the expected outcomes of these operations?

```
cudf.DataFrame({'a':[100], 'b':[None]}).query("a >= 0 or b <= 0)
```
right now I'm filtering out this row, because one of the conditions _would_ evaluate to `null`. But I'm uneasy about this because I'm not sure if we should treat `or` as a bitwise operation, which we'd be doing if we merged this PR. If we treat is like a bitwise operator, we don't get any rows back. If we treat it like a logical operator, would we? And if so, which one is desirable?

**Comment by shwina** [comment]
`and` and `or` should be considered logical operators in `query` - from the Pandas `query` docs:

> For example, the & and | (bitwise) operators have the precedence of their boolean cousins, and and or.

Here's our current behaviour for logical `or`:

print(cudf.logical_or(cudf.Series([None], dtype='int32'), cudf.Series([100], dtype='int32')))                                                                                                          
0     
dtype: bool

To truly match Pandas behaviour:

print(cudf.logical_or(cudf.Series([np.nan], nan_as_null=False), cudf.Series([100])))                                                                                                                   
0    True
dtype: bool

**Comment by cwharris** [comment]
After discussing with @kkraus14 , pessimistic null handling seems good enough.

File: python/cudf/cudf/dataframe/dataframe.py [hunk 1]
```diff
@@ -2556,6 +2556,8 @@ def query(self, expr, local_dict={}):
 
             Names starting with `@` refer to Python variables
 
+            A row will be excluded if it provides a null value to expr
```
Comment (kkraus14) [inline python/cudf/cudf/dataframe/dataframe.py]:
```suggestion
            An output value will be `null` if any of the input values are
            `null` regardless of expression
```

**Pull request merged**